quivr/tables.py

```diff
@@ -1082,3 +1085,44 @@ def set_column(self, name: str, data: DataSourceType) -> Self:
 
         table = column._set_on_pyarrow_table(self.table, data)
         return self.from_pyarrow(table=table, validate=True, permit_nulls=False)
+
+    def parallel_apply(
+        self,
+        func: Callable[[Self], T],
+        chunk_size: int = 10000,
+        max_workers: Optional[int] = None,
+    ) -> Iterator[T]:
+        """Call func on self's data in parallel.
+
+        :param func: A function that takes a Table and returns a value.
+        :param chunk_size: The number of rows to process in each
+        chunk. One chunk may be smaller than this if the table has
+        fewer rows.
+
+        :param max_workers: The maximum number of threads to use. If
+        None, the number of threads is chosen automatically to be the
+        number of CPU cores.
+
+        :return: An iterable of the return values of func. The order of
+        the values is not guaranteed to match the order of the rows in
+        the table.
+
+        """
+        if max_workers is None:
+            max_workers = os.cpu_count()
+
+        futures = []
+
+        def actual_func(batch: pa.RecordBatch) -> T:
+            table = pa.Table.from_batches([batch], schema=self.table.schema)
+            # Assume the table remains valid for batches
+            instance = self.from_pyarrow(table, validate=False)
```

Comment:
I'm assuming this is basically zero copy with only the tiny overhead of the quivr table instance pointing to the batch.